runners/direct-java/src/main/java/org/apache/beam/runners/direct/ExecutorServiceParallelExecutor.java

```diff
@@ -455,17 +464,55 @@ private boolean shouldShutdown() {
      * add more work from root nodes that may have additional work. This ensures that if a pipeline
      * has elements available from the root nodes it will add those elements when necessary.
      */
-    private void addWorkIfNecessary(boolean firedTimers) {
+    private void addWorkIfNecessary() {
       // If any timers have fired, they will add more work; We don't need to add more
-      if (firedTimers) {
-        return;
-      }
-      // All current TransformExecutors are blocked; add more work from the roots.
-      for (AppliedPTransform<?, ?, ?> root : rootNodes) {
-        if (!evaluationContext.isDone(root)) {
-          scheduleConsumption(root, null, defaultCompletionCallback);
+      if (state.get() == ExecutorState.QUIESCENT) {
+        // All current TransformExecutors are blocked; add more work from the roots.
+        for (AppliedPTransform<?, ?, ?> root : rootNodes) {
+          if (!evaluationContext.isDone(root)) {
+            scheduleConsumption(root, null, defaultCompletionCallback);
+            state.set(ExecutorState.ACTIVE);
+          }
         }
       }
     }
   }
+
+
+  /**
+   * The state of the executor. The state of the executor determines the behavior of the
+   * {@link MonitorRunnable} when it runs.
+   */
+  private enum ExecutorState {
+    /**
+     * The Executor has new work available to it. All pending, including potentially blocked work,
+     * should be evaluated.
+     *
+     * <p>The executor becomes active whenever a timer fires, a {@link PCollectionView} is updated,
+     * or output is produced by the evaluation of a {@link TransformExecutor}.
+     */
+    ACTIVE,
+    /**
+     * The Executor does not have any new work available to it, but work is in progress. New work
+     * should not be added until the Executor becomes active or no work is pending.
+     *
+     * <p>If all outstanding work completes without the executor becoming {@code ACTIVE}, the
+     * Executor enters state {@code QUIESCING}. Pending work must be reevaluated.
+     */
+    PROCESSING,
+    /**
+     * The Executor had no new work and no outstanding work. All outstanding work is work that
```

Comment:
This description is phrased in past tense, but a state should describe the current state of affairs.

Possibly related to this it says it "had... no outstanding work" but then "All outstanding work is work that may be blocked on a side input". One or the other of these descriptions needs revision.